src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs

```diff
@@ -5375,6 +5375,117 @@ .maxstack  3
   IL_004f:  box        ""int""
   IL_0054:  stfld      ""object C.<>c__DisplayClass0_0.o""
   IL_0059:  ret
+}");
+        }
+
+        [WorkItem(44720, "https://github.com/dotnet/roslyn/issues/44720")]
+        [Fact]
+        public void LambdaInsideGenericLocalFunctionInsideLoop_CompilesCorrectly()
+        {
+            // The lambda passed into StaticMethod depends on TLocal type argument of LocalMethod
+            // so if the delegate is cached outside the method the type argument reference is broken
+            // and code throws BadImageFormatException. Such a broken code will looks like:
+            // class DisplayClass
+            // {
+            //     Func<TLocal, string> _cachedDelegate;
+            //
+            //     void LocalMethod<TLocal>()
+            //     {
+            //         ...
+            //     }
+            //
+            // The test checks that it is not the issue.
+            string source =
+                @"using System;
+using System.Collections.Generic;
+
+static class Program
+{
+    private static void Main()
+    {
+        TestMethod(string.Empty);
+    }
+
+    private static void TestMethod<T>(T param)
+    {
+        var message = string.Empty;
+
+        for (int i = 0; i < 1; i++)
+        {
+            void LocalMethod<TLocal>(TLocal value)
```

Comment:
>void LocalMethod<TLocal>(TLocal value) [](start = 12, length = 38)

I think we should test more permutations and whether caching is used for them:
- both the lambda and the loop are inside a local function 
- This local function is not generic and is inside another local function, the used type parameter belongs to the outer local function rather than to this one,   
- there is no local function, the delegate refers to the type parameter of the enclosing method
- the type parameter of the local function is not used in the delegate type
- etc.

Obviously, if you plan to stick with new ```HasTypeArgumentsFromReferencedMethod``` method, we would want to test more delegate types to cover full breadth of different type symbol kinds.